dbms/src/Interpreters/InterpreterSelectWithUnionQuery.cpp

```diff
@@ -196,14 +197,19 @@ BlockInputStreams InterpreterSelectWithUnionQuery::executeWithMultipleStreams(Qu
         parent_pipeline.addInterpreterContext(context);
     }
 
+    /// Update max_streams due to:
+    /// - max_distributed_connections for Distributed() engine
+    /// - max_streams_to_max_threads_ratio
+    max_streams = (*std::min_element(nested_interpreters.begin(), nested_interpreters.end(), [](const auto &a, const auto &b) {
```

Comment:
I suppose using minimum is not the best choice. Imagine query like `select number from system.number union all select number from system.numbers`. It could have been executed in in 2 threads, but after this change it it will use only one.

I think we should think more about it.

Reply:
Indeed (I also though like this), and this will go away, already testing better approach

Reply:
Hm, I though that I found better alternative but I didn't (too much interpreters:) )

As for `union all`:
- if `max_threads=1` will be specified than it should use one thread (like before)
- if `max_threads=2` will be specified than it will be runned in parallel (like before, since each interpreter will have `max_streams=2`)

@KochetovNicolai what did I miss?

Reply:
FWIW:

```sh
$ time clickhouse-client --max_threads 1 -q 'select sleep(1) union all select sleep(1)'
0
0
real    0m2.322s
$ time clickhouse-client --max_threads 2 -q 'select sleep(1) union all select sleep(1)'
0
0
real    0m1.327s
$ time clickhouse-client --max_threads 2 -q 'select sleep(1) union all select sleep(1)' --experimental_use_processors 0 
0
0
real    0m1.419s
$ time clickhouse-client --max_threads 1 -q 'select sleep(1) union all select sleep(1)' --experimental_use_processors 0 
0
0
real    0m2.421s
```